Documentation updates (#15)
neptune-ai/neptune-client-scale: README.md

Reviewer: Let's mention that Windows is not currently supported

Reviewer 2: Isin't it?

# Neptune Scale client

> [!NOTE]

> > This package only works with the `3.0` version of neptune.ai called Neptune Scale, which is in beta.

Author: ```suggestion
> This package only works with the `3.0` version of neptune.ai called Neptune Scale, which is in beta.
>
> It's supported on Linux and MacOS.
```

Reviewer: Can we have a separate _Prerequisites_ section instead, listing Neptune Scale, Python 3.8+, and Linux/MacOS platforms?
Until we have the docs out, we can use this section to keep adding any other prerequisites

>
> You can't use the Scale client with the stable Neptune `2.x` versions currently available to SaaS and self-hosting customers. For the Python client corresponding to Neptune `2.x`, see https://github.com/neptune-ai/neptune-client.

**What is Neptune?**

Neptune is an experiment tracker. It enables researchers to monitor their model training, visualize and compare model metadata, and collaborate on AI/ML projects within a team.

**What's different about Neptune Scale?**

Neptune Scale is the next major version of Neptune. It's built on an entirely new architecture for ingesting and rendering data, with a focus on responsiveness and accuracy at scale.

Neptune Scale supports forked experiments, with built-in mechanics for retaining run ancestry. This way, you can focus on analyzing the latest runs, but also visualize the full history of your experiments.

## Installation

```bash

> pip install neptune-client-scale

Author: By the time these instructions are ready to be followed, the package should be available on PyPI, no?

```suggestion
pip install neptune-scale
```

Reviewer 2: Yes, it works right now. Go ahead and apply the suggestion

> pip install neptune-client-scale

Reviewer 2: Closing as outdated. `pip install neptune-client-scale`

> pip install neptune-client-scale

Reviewer 2: Right now it's `neptune-client-scale` but it's going to be renamed (probably today EOD)

> pip install neptune-client-scale

Reviewer: Lets remember to remove `neptune-client-scale` from PyPI in that case

> pip install neptune-client-scale

Reviewer 2: Done

| `resume`         | `bool`, optional | `False` | If `False` (default), creates a new run. To continue an existing run, set to `True` and pass the ID of an existing run to the `run_id` argument. To fork a run, use `from_run_id` and `from_step` instead. |
| `mode`           | `Literal`, `"async"` or `"disabled"` | `"async"` | Mode of operation. If set to `"disabled"`, the run doesn't log any metadata. |
| `as_experiment`  | `str`, optional  | `None` | Name of the experiment to associate the run with. Learn more about [experiments](https://docs-beta.neptune.ai/concepts) in the Neptune documentation. |
| `creation_time`  | `datetime`, optional | `None` | Custom creation time of the run. |
| `from_run_id`    | `str`, optional  | `None` | If forking off an existing run, ID of the run to fork from. |
| `from_step`      | `int`, optional  | `None` | If forking off an existing run, step number to fork from. |
| `max_queue_size` | `int`, optional  | 1M | Maximum number of operations queued for processing. 1 000 000 by default. You should raise this value if you see the `on_queue_full_callback` function being called. |

> | `on_queue_full_callback` | `Callable[[BaseException, Optional[float]], None]`, optional | `None` | Callback function triggered when the queue is full. The function should take two arguments: (1) Exception that made the queue full. (2) Optional timestamp: When the exception was last raised. |

Author: It doesn't seem ideal to use "must" with optional args... But we can rephrase:

```suggestion
| `on_queue_full_callback` | `Callable[[BaseException, Optional[float]], None]`, optional | `None` | Callback function triggered when the queue is full. The function must take as an argument the exception that made the queue full and, as an optional argument, a timestamp of when the exception was last raised. |
```

This is a blocking operation. Call the function at the end of your script, after your model training is completed.

__Examples__

```python
from neptune_scale import Run

run = Run(...)
run.log(...)

run.close()
```

If using a context manager, Neptune automatically closes the run upon exiting the context:

```python
with Run(...) as run:
    ...

# run is closed at the end of the context
```

## `log()`

Logs the specified metadata to a Neptune run.

You can log metrics, tags, and configurations. Pass the metadata as a dictionary `{key: value}` with

- `key`: path to where the metadata should be stored in the run.
- `value`: the piece of metadata to log.

For example, `{"parameters/learning_rate": 0.001}`. In the field path, each forward slash `/` nests the field under a namespace. Use namespaces to structure the metadata into meaningful categories.

__Parameters__

| Name          | Type                                               | Default | Description                                                               |
|---------------|----------------------------------------------------|---------|---------------------------------------------------------------------------|

> | `step`        | `Union[float, int]`, optional                      | `None`  | Index of the log entry. Must be increasing. If not specified, the `log()` call increments the step starting from the highest already logged value. **Tip:** Using float rather than int values can be useful, for example, when logging substeps in a batch. |

Author: Then how does this work when omitting the step? Or are we missing that this is connected to metrics (FloatSeries) only?

| `timestamp`   | `datetime`, optional                               | `None`  | Time of logging the metadata. |

> | `fields`      | `Dict[str, Union[float, bool, int, str, datetime, list, set]]`, optional  | `None` | Dictionary of configs or other values to log. Independent of the step value. Available types: float, integer, Boolean, string, and datetime. To log multiple values at once, pass multiple dictionaries. |

Author: I think it's just an oversight on my part. Should be a single dictionary with multiple key-value pairs, right?

> | `fields`      | `Dict[str, Union[float, bool, int, str, datetime, list, set]]`, optional  | `None` | Dictionary of configs or other values to log. Independent of the step value. Available types: float, integer, Boolean, string, and datetime. To log multiple values at once, pass multiple dictionaries. |

Author: ```suggestion
| `fields`      | `Dict[str, Union[float, bool, int, str, datetime, list, set]]`, optional  | `None` | Dictionary of configs or other values to log. Available types: float, integer, Boolean, string, and datetime. |
```

| `add_tags`    | `Dict[str, Union[List[str], Set[str]]]`, optional  | `None`  | Dictionary of tags to add to the run, as a list of strings. Independent of the step value. |
| `remove_tags` | `Dict[str, Union[List[str], Set[str]]]`, optional  | `None`  | Dictionary of tags to remove from the run, as a list of strings. Independent of the step value. |

__Examples__

Create a run and log some metadata:

```python
from neptune_scale import Run

with Run(...) as run:
    run.log(
        fields={"parameters/learning_rate": 0.001},
        add_tags={"sys/tags": ["tag1", "tag2"]},
        metrics={"loss": 0.14, "acc": 0.78},
    )
```

Remove a tag:

```python
with Run(...) as run:
    run.log(remove_tags={"sys/tags": "tag2"})
```

You can pass the step when logging metrics:

```python
run.log(step=5, metrics={"loss": 0.09, "acc": 0.82})  # works if the previous step is no higher than 4
run.log(metrics={"loss": 0.08, "acc": 0.86})          # step index is set to "6"
...
```

> **Note:** Calling `log()` without specifying the step still increments the index. To correlate logged values, make sure to send all metadata related to a step in a single `log()` call, or specify the step explicitly.

Author: Yes, but I'm not sure it helps clarify which part of the note is false. Maybe the problem is that I didn't include a mention of `metrics`?

Isn't it true that if the `step` is omitted when calling `log_metrics()`, the highest index found among logged FloatSeries fields is used as the reference step?

# Neptune Scale client

> [!NOTE]
> This package only works with the `3.0` version of neptune.ai called Neptune Scale, which is in beta.
>

> > It's supported on Linux and MacOS.

Reviewer: Update to this.
It can be used on Windows if the run is initialized inside the `if __name__ == "__main__":` guard.

Read more here: https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming > _Safe importing of main module_

cc: @Raalsky

Reviewer 2: It's even not specific to Windows, it's common for all of the operating systems and a common pitfall of Python itself.

Reviewer: I was able to use it without the guard in Linux 🤔

Reviewer 2: Weird, on Mac it doesn't work without the guard.

Author: I conclude that we don't need any OS-specific notes for now.

Reviewer: To cover our bases, I'd prefer if we add a debugging/Help section of sorts and inform users to initialize the run inside  the `if __name__ == "__main__":` guard if they get the below error:

```
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.
```

>
> You can't use the Scale client with the stable Neptune `2.x` versions currently available to SaaS and self-hosting customers. For the Python client corresponding to Neptune `2.x`, see https://github.com/neptune-ai/neptune-client.

**What is Neptune?**

Neptune is an experiment tracker. It enables researchers to monitor their model training, visualize and compare model metadata, and collaborate on AI/ML projects within a team.

**What's different about Neptune Scale?**

Neptune Scale is the next major version of Neptune. It's built on an entirely new architecture for ingesting and rendering data, with a focus on responsiveness and accuracy at scale.

Neptune Scale supports forked experiments, with built-in mechanics for retaining run ancestry. This way, you can focus on analyzing the latest runs, but also visualize the full history of your experiments.

## Installation

```bash
pip install neptune-scale
```

## Example usage

```python
from neptune_scale import Run

run = Run(
    family="RunFamilyName",
    run_id="SomeUniqueRunIdentifier",
)

run.log(

>     metrics={"Metric1": metric1_value, "Metric2": metric2_value},

Reviewer: ```suggestion
    metrics={"acc": 0.98, "loss": 0.2},
```

>     fields={"Field1": field1_value}

Reviewer: ```suggestion
    fields={"params/lr": 0.01, "params/optimizer": "adam"}
```

)

run.close()
```

## API reference

### `Run`

Representation of experiment tracking metadata logged with Neptune Scale.

#### Initialization

Initialize with the class constructor:

```python
from neptune_scale import Run

run = Run(...)
```

or using a context manager:

```python
from neptune_scale import Run

with Run(...) as run:
    ...
```

__Parameters__

| Name             | Type             | Default | Description                                                               |
|------------------|------------------|---------|---------------------------------------------------------------------------|

> | `family`         | `str`            | -       | Identifies related runs. All runs of the same lineage must have the same `family` value. That is, forking is only possible within the same family. Max length: 128 characters. |

Reviewer: ```suggestion
| `family`         | `str`            | -       | Identifies related runs. All runs of the same lineage must have the same `family` value. That is, forking is only possible within the same family. Max length: 128 bytes. |
```

> | `run_id`         | `str`            | -       | Identifier of the run. Must be unique within the project. Max length: 128 characters. |

Reviewer: ```suggestion
| `run_id`         | `str`            | -       | Identifier of the run. Must be unique within the project. Max length: 128 bytes. |
```

| `project`        | `str`, optional  | `None`  | Name of a project in the form `workspace-name/project-name`. If `None`, the value of the `NEPTUNE_PROJECT` environment variable is used. |
| `api_token`      | `str`, optional  | `None`  | Your Neptune API token or a service account's API token. If `None`, the value of the `NEPTUNE_API_TOKEN` environment variable is used. To keep your token secure, don't place it in source code. Instead, save it as an environment variable. |
| `resume`         | `bool`, optional | `False` | If `False` (default), creates a new run. To continue an existing run, set to `True` and pass the ID of an existing run to the `run_id` argument. To fork a run, use `from_run_id` and `from_step` instead. |
| `mode`           | `Literal`, `"async"` or `"disabled"` | `"async"` | Mode of operation. If set to `"disabled"`, the run doesn't log any metadata. |

> | `as_experiment`  | `str`, optional  | `None` | Name of the experiment to associate the run with. Learn more about [experiments](https://docs-beta.neptune.ai/experiments) in the Neptune documentation. |

Reviewer: ```suggestion
| `as_experiment`  | `str`, optional  | `None` | Name of the experiment to associate the run with. Learn more about [experiments](https://docs-beta.neptune.ai/experiments) in the Neptune documentation. Max length: 730 bytes |
```
